src/test/pcp_tests.cpp

```diff
@@ -0,0 +1,701 @@
+
+/// UDP test server operation.
+struct TestOp {
+    std::chrono::milliseconds delay;
+    enum Op {
+        SEND, // Expect send (with optional data)
+        RECV, // Expect receive (with data)
+        NOP,  // Just delay
+    } op;
+    std::vector<uint8_t> data;
+
+    //! Injected error.
+    //! Set this field to a non-zero value to return the errno code from send or receive operation.
+    int error;
+
+    TestOp(std::chrono::milliseconds delay_in, Op op_in, const std::vector<uint8_t> &data_in, int error_in):
+        delay(delay_in), op(op_in), data(data_in), error(error_in) {}
+};
+
+/// Save the value of CreateSock and restore when the test ends.
+class PCPTestingSetup : public BasicTestingSetup
+{
+public:
+    explicit PCPTestingSetup(const ChainType chainType = ChainType::MAIN,
+                             TestOpts opts = {})
+        : BasicTestingSetup{chainType, opts},
+          m_create_sock_orig{CreateSock}
+    {
+        const std::optional<CService> local_ipv4{Lookup("192.168.0.6", 1, false)};
+        const std::optional<CService> local_ipv6{Lookup("2a10:1234:5678:9abc:def0:1234:5678:9abc", 1, false)};
+        const std::optional<CService> gateway_ipv4{Lookup("192.168.0.1", 1, false)};
+        const std::optional<CService> gateway_ipv6{Lookup("2a10:1234:5678:9abc:def0:0000:0000:0000", 1, false)};
+        BOOST_REQUIRE(local_ipv4 && local_ipv6 && gateway_ipv4 && gateway_ipv6);
+        default_local_ipv4 = *local_ipv4;
+        default_local_ipv6 = *local_ipv6;
+        default_gateway_ipv4 = *gateway_ipv4;
+        default_gateway_ipv6 = *gateway_ipv6;
+
+        struct in_addr inaddr_any;
+        inaddr_any.s_addr = htonl(INADDR_ANY);
+        bind_any_ipv4 = CNetAddr(inaddr_any);
+    }
+
+    ~PCPTestingSetup()
+    {
+        CreateSock = m_create_sock_orig;
+        MockableSteadyClock::ClearMockTime();
+    }
+
+    // Default testing nonce.
+    static constexpr PCPMappingNonce TEST_NONCE{0x11, 0x22, 0x33, 0x44, 0x55, 0x66, 0x77, 0x88, 0x99, 0xaa, 0xbb, 0xcc};
+    // Default network addresses.
+    CNetAddr default_local_ipv4;
+    CNetAddr default_local_ipv6;
+    CNetAddr default_gateway_ipv4;
+    CNetAddr default_gateway_ipv6;
+    // IPv4 bind
+    CNetAddr bind_any_ipv4;
+private:
+    const decltype(CreateSock) m_create_sock_orig;
+};
+
+/** Simple scripted UDP server emulation for testing.
+ */
+class PCPTestSock final : public Sock
+{
+public:
+    // Note: we awkwardly mark all methods as const, and properties as mutable,
+    // because Sock expects all networking calls to be const.
+    explicit PCPTestSock(const CNetAddr &local_ip, const CNetAddr &gateway_ip, const std::vector<TestOp> &script)
+        : Sock{INVALID_SOCKET},
+          m_script(script),
+          m_local_ip(local_ip),
+          m_gateway_ip(gateway_ip)
+    {
+        ElapseTime(std::chrono::seconds(0)); // start mocking steady time
+        PrepareOp();
+    }
+
+    ~PCPTestSock() override { }
```

Comment:
clang-tidy complains about this:
> [09:51:32.084] [584/666][11.5s] clang-tidy-19 -p=/ci_container_base/ci/scratch/build-x86_64-pc-linux-gnu -quiet -load=/tidy-build/libbitcoin-tidy.so /ci_container_base/src/test/pcp_tests.cpp
[09:51:32.084] /ci_container_base/src/test/pcp_tests.cpp:96:5: error: use '= default' to define a trivial destructor [modernize-use-equals-default,-warnings-as-errors]
[09:51:32.084]    96 |     ~PCPTestSock() override { }
[09:51:32.084]       |     ^                       ~~~
[09:51:32.084]       |                             = default;
[09:51:32.084] 973 warnings generated.

Reply:
Yes, this is really silly. Linters are useful for preventing constructs that have high risk of bugs but this just seems like busywork.
Anyhow, removed the destructor completely and re-pushed.

Reply:
Nah it's not a big deal i should probably find a way to run the linter stuff locally to prevent having to play ping pong like this.